var/spack/repos/builtin/packages/hypre/package.py

```diff
@@ -220,26 +220,50 @@ def install(self, spec, prefix):
                         '-rhsone')
             make("install")
 
+    extra_install_tests = join_path('src', 'examples')
+
     @run_after('install')
     def cache_test_sources(self):
-        srcs = ['src/examples']
-        self.cache_extra_test_sources(srcs)
+        self.cache_extra_test_sources(self.extra_install_tests)
+
+    @property
+    def _cached_tests_work_dir(self):
+        """The working directory for cached test sources."""
+        return join_path(self.test_suite.current_test_cache_dir,
+                         self.extra_install_tests)
+
+    def _test_build_examples(self):
+        """Build cached test examples."""
+        return self.run_test('make',
+                             ['HYPRE_DIR={0}'.format(self.prefix), 'bigint'],
+                             purpose='test: building selected examples',
+                             work_dir=self._cached_tests_work_dir)
+
+    def _test_clean_examples(self):
+        """Clean up any test example build files."""
+        return self.run_test('make', ['distclean'], [],
+                             purpose='test: removing prior test example build',
+                             work_dir=self._cached_tests_work_dir)
```

Comment:
Is it necessary to clean the build directory? I thought the test stage was unique for every invocation of `spack test run`, so that spack copies the test files to an empty stage directory.

Reply:
The hash for a given installed spec will be unchanged so, if you run `spack test run --keep-stage` for the same spec(s), then the test stage will be retained.  I'm *pretty sure* the directory doesn't actually get replaced but I need to check to be sure.

Reply:
Ah I see, I assumed that since the test hash was different from the spec hash that it was unique for a particular test run. FWIW I think in general it would be a better (at the very least, more consistent) approach for spack to clean the test directory before running the tests. After all, when you rebuild a package with spack, even when you tell it to keep the previous stage, it'll forcibly clean it before trying to build again. This also will help reduce the complexity of the smoke test recipes since cleaning will be done in a single place.

Reply:
> Ah I see, I assumed that since the test hash was different from the spec hash that it was unique for a particular test run. FWIW I think in general it would be a better (at the very least, more consistent) approach for spack to clean the test directory before running the tests. After all, when you rebuild a package with spack, even when you tell it to keep the previous stage, it'll forcibly clean it before trying to build again. This also will help reduce the complexity of the smoke test recipes since cleaning will be done in a single place.

FWIW. The builds for `hypre` and `openmpi` are written such that the binaries are updated on every build even when re-running after a `--keep-stage`.  So I removed the clean operations.

I also removed the the conditional skip if the build failed.

As for Spack automatically cleaning tests being re-run in a `--keep-stage` directory, I've brought this up with becker33.

A feature in such cases is that the test results (summary) accumulates results from subsequent runs so any clean operation has to avoid wiping that data to retain that information.  

There's also the matter that the clean process for examples/tests will depend on the package build process so a general solution would be to wipe out and re-copy files in this case.  Perhaps that's a non-issue.